src/allocator.rs

```diff
@@ -332,13 +336,36 @@ impl Allocator {
 
     pub fn new_pair(&mut self, first: NodePtr, rest: NodePtr) -> Result<NodePtr, EvalErr> {
         let idx = self.pair_vec.len();
-        if idx == MAX_NUM_PAIRS {
+        if idx >= self.max_num_pairs {
             return err(self.nil(), "too many pairs");
         }
         self.pair_vec.push(IntPair { first, rest });
         Ok(NodePtr::new(ObjectType::Pair, idx))
     }
 
+    pub fn reduce_pair_max(&mut self, amount: usize) -> Result<(), EvalErr> {
+        if self.max_num_pairs - amount <= self.pair_vec.len() {
+            return err(
+                self.nil(),
+                format!(
+                    "reduce_pair_max: amount exceeds current pair count ({} pairs allocated, {} max pairs)",
+                    self.pair_vec.len(),
+                    self.max_num_pairs
+                ).as_str(),
```

Comment:
I think it would be nice to have the same error message as `new_pair()` here. a string literal "too many pairs".

Reply:
Done